src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,352 @@ pub mod policy;
+            accumulated_constraints.accumulate_san(working_cert_extensions)?;
+        }
+
+        accumulated_constraints.apply_and_accumulate(working_cert_extensions)?;
+
+        // Look in the store's root set to see if the working cert is listed.
+        // If it is, we've reached the end.
+        if self.store.contains(working_cert) {
+            return Ok(vec![working_cert.clone()]);
+        }
+
+        // Otherwise, we collect a list of potential issuers for this cert,
+        // and continue with the first that verifies.
+        let mut last_err: Option<ValidationError> = None;
+        for issuing_cert_candidate in self.potential_issuers(working_cert) {
+            // A candidate issuer is said to verify if it both
+            // signs for the working certificate and conforms to the
+            // policy.
+            let issuer_extensions = issuing_cert_candidate.extensions()?;
+            match self.policy.valid_issuer(
+                issuing_cert_candidate,
+                working_cert,
+                current_depth,
+                &issuer_extensions,
+            ) {
+                Ok(next_depth) => {
+                    match self.build_chain_inner(
+                        issuing_cert_candidate,
+                        next_depth,
+                        false,
+                        &issuer_extensions,
+                        accumulated_constraints,
+                    ) {
+                        Ok(mut chain) => {
+                            chain.insert(0, working_cert.clone());
+                            return Ok(chain);
+                        }
+                        Err(e) => last_err = Some(e),
+                    };
+                }
+                Err(e) => last_err = Some(e),
+            };
+        }
+
+        // We only reach this if we fail to hit our base case above, or if
+        // a chain building step fails to find a next valid certificate.
+        Err(ValidationError::CandidatesExhausted(last_err.map_or_else(
+            || {
+                Box::new(ValidationError::Other(
+                    "all candidates exhausted with no interior errors".to_string(),
+                ))
+            },
+            |e| match e {
+                // Avoid spamming the user with nested `CandidatesExhausted` errors.
+                ValidationError::CandidatesExhausted(e) => e,
+                _ => Box::new(e),
+            },
+        )))
+    }
+
+    fn build_chain(&self, leaf: &'a Certificate<'chain>) -> Result<Chain<'chain>, ValidationError> {
+        // Before anything else, check whether the given leaf cert
+        // is well-formed according to our policy (and its underlying
+        // certificate profile).
+        //
+        // In the case that the leaf is an EE, this includes a check
+        // against the EE cert's SANs.
+        let leaf_extensions = leaf.extensions()?;
+
+        self.policy.permits_leaf(leaf, &leaf_extensions)?;
```

Comment:
I thought about this a bit more, and I think this is currently slightly wrong for two reasons:

1. The MVP is Web PKI compatibility, which forbids CAs in the leaf position (the leaf is always a subscriber, and subscriber certificates must not be CA certificates)
2. In the current implementation, when the leaf is a CA, we don't end up checking its SANs (since we currently only enforce SANs on EEs)

(2) is probably just wrong, but fixing it is a little weird: we'll need some kind of `ca::subject_alternative_name` extension validator, which will then need to be aware whether the cert is in leaf position or now (since the SAN should otherwise be ignored). It might be easier to fix (1) instead, which would then eliminate (2) transitively. This is what `rustls` and `webpki` appear to do.

Reply:
So: a) this suggests a missing test case in x509-limbo, the case of an EE being a CA without a SAN for the requested name, b) it's a problem because it means that an attacker who passes an intermediate as their EE gets to be valid for all names.

To me the fix is that `permits_leaf` needs to check the SAN, it's not about EE vs. CA.